_posts/2019-04-24-istio-service-mesh-s2s.md

```diff
@@ -0,0 +1,428 @@
+category: Cloud
+comments: true
+---
+
+> This post will describe how to use the Istio service mesh to provide service to service authentication and authorization in a Kubernetes cluster.
+> It will show how ServiceRoles, ServiceRoleBindings and Identities in Istio can be used to achieve this. 
+
+# Table of content 
+
+* [What is Istio?](#what-is-istio)
+* [Istio concepts](#istio-concepts)
+* [Show me the code](#show-me-the-code)
+* [Conclusion](#conclusion)
+
+## What is Istio?
+
+Istio is a service mesh created by Google, Lyft and IBM. 
+It aims to simplify some security and management aspects of a microservices software architecture.
+More information on Istio and its feature can be found in its [docs](https://istio.io/docs/){:target="_blank" rel="noopener noreferrer"}.
+In this blogpost we will highlight one of the key security features of Istio: service to service authentication and authorization.
+For the sake of simplicity, this post will focus on an Istio setup in Kubernetes.
+
+In a microservices architecture, managing access to services can be a challenging operation. 
+For end-user facing services, JWTs are used to add authenication information a request.
+They are used by the service to determine what end-user is making the request.
+These tokens can be generated based on information that the end-user provides to an identity provider.
+In most cases this information is a username and password, with some additional 2FA if possible.
+This setup can be achieved by using OpenId Connect as a protocol with the authorization code grant flow and an identity provider like Keycloak.
+
+When services communicate with each other, they also need to provide an identity to each other.
+A common option to do this is using by using client credentials grant flow of OpenId Connect.
+In this flow, a service provides a client credentials to the identity provider to generate a token.
+This token will be used to communicate to a service.
+
+Both of these types of authentication flows are application level authentication flows. 
+They allow services to determine what resources an end-user or service can access.
+Istios service to service role based acccess control (RBAC) is not on an application level but on an communication level.
+It specifies which services can connect and communicate with each other. 
+In order to achieve this, Istio connects an identity to each service in the mesh and allows it to authenticate itself.
+The requested service can use this identity to determine if the service is allowed to connect or not.  
+Istio makes use of proxies to handle all traffic (into and out of services) and using mutual trusted certificates to secure the connection and provide an identity to these proxies.
+When using the automatic proxy injection, enabling Istio's service to service RBAC mechanism is almost as easy as flipping a switch.
+
+There are five main components responsible for making this possible in Istio: Citadel, Pilot, Galley, Mixer and Envoy. 
```

Comment:
Maybe emphasize the components in the next paragraphs by making it **bold**?

Reply:
Updated